### PR TITLE
Add utility function for expanding data

### LINKED ISSUE #1366: Misalignment in graphs (and reports) due to instantiation of new variables during simulation
## Status
This is a problem encountered in #1263, specific to the creation of new pens, and it may extend to other outputs.
When a new output is added to output manager, it causes misalignment in graphs and other outputs, because an assumption that the first entry is on simulation day 0 is incorrect.

**A super-simplified example of the problem**
Assume one runs a simulation for 15 days.
The simulation starts with 4 pens.
A 5th pen is instantiated on day 10 (see logic in #1263).
At the end of the simulation, each output related to pens 0-3 have a length of 15 (or 16, if we're counting day 0).

*The problem:* any outputs related to pen 4 has 10 fewer days, and so there's a misalignment in those outputs.

**Example output from a longer sim**
Note the crude arrows showing shorter line lengths.
default_animalmodulereporter-daily_pens-05-Mar-2024_Tue_11-10-38

Note below the hidden rows to show the "ragged" nature of these new pen outputs (this is without the "padding" I instituted in 1263).
image

So without some note of what the time is at the time of collection, we lose information. (Note in "Yet another solution" that we can already store that in info_map, but that's harder for an end user to access, and lost completely in graphs).

## Intention
We need to ensure that the user is certain of when data was logged in output manager.

**What the above *should* look like** (thanks @Isgotkowitz for the suggestion to show some examples of ideal behavior)
- There are two options I see for the reports/CSVs:
    - Each variable would have its x-axis value printed next to it, so that there would be double the number of columns in that example. 
image

    - To avoid the duplication inherent in the previous bullet point, there could be a single x-axis that all results are aligned with, e.g. the first column is simulation_day, and every entry in a column is aligned to that axis.
    - However, this would have a problem when printing a mix of monthly and daily variables, since there would be enormous gaps, and would be annoying for those mainly working in Excel.
image

- As for graphs, the end of each of the lines on the graph would all align to the x-axis (in the above examples, they'd align at the end). Then the user can see when the new pens "start" existing. This example is harder to competently mock up, but hopefully this one is easy to imagine.

**Additional problems**
- If a pen (or other outputs) are intermittently recorded, those gaps are also problematic. After chatting with @Isgotkowitz there may be a similar issue in crop and soil, e.g. if corn is grown on days 150-220 of the simulation, 

**Current implementation**
In that PR I (@JoeWaddell) added some janky code that pads out calls to OM to fill in gaps with blanks (still only local). However, this requires an additional set of 5+ lines for each call to output manager, and requires that any new pen-specific outputs are carefully managed.

**Current workaround**
This doesn't solve the problem re: misalignment in graphs, but for reports, when slicing from the last entries, those misalignments are avoided (provided that they aren't reported intermittently).

## Request: in a sentence
Record simulation day alongside every call to the OM variables pool.

**Proposed solution**
Like the new units field, adding simulation_day could be an option when adding variables to output manager. Sometime this addition might not be necessary, so it could also be an optional arg.

**Another solution**
OM could automatically log when in the simulation a variable was added, so it would require no changes to currently implemented calls to OM.

**Yet another solution** Graph and report generators could be enhanced to look for `simulation_day` in `info_map` (if available) to set the x-axis variable.

## Additional thoughts

**Additional benefits**
- Monthly and daily outputs could be plotted on the same figure, since the x-axis would be defined explicitly.
- Multiple points can show up for a single day: e.g. if we're recording individual animal data.
- Would work nicely with #912, since it's an assumption there that the data on the x-axis are aligned such that the first entry is day 0.

**Complications**
- This would require not only changes to OM, but graph generator, report generator, etc, which may be easier said than done.
- This also means more information is being logged to OM. Granted, only an integer, but still: one for each variable. Consideration should be given to a potential increase in file size and simulation time, and whether all of this is particularly useful now or in the future.

### PR DESCRIPTION
**Pull request by Isgotkowitz**
This PR lays the groundwork for padding and aligning data temporally in graphs and reports.

This PR will be followed up with a PR that incorporates the padding capabilities into the Graph Generator and Report Generator. - PR #1778 

### Context

Issue(s) closed by this pull request: addresses #1366 

- [x] The [changelog](https://docs.google.com/spreadsheets/d/1U-3igxdstHTFb6k5dVcE5-x9IL0r6q18eUiwwytKZyA/edit#gid=0) has been updated.
      
How big are the changes in the PR? Would you nominate it for a major version upgrade?
- [ ] Yes
- [x] No

### What

In order to align values from different variables, the simulated day that a variable is recorded on is now automatically added to the info map, though some options are provided to override this behavior. A new utility method has been introduced so that it can be 

### Why

The "when" of data is very important, and if that can't be captured in outputs then users will lose key information which will limit interpretation of results. #1366 gets into the why as well.

### How

The Output Manager's `add_variable` and `_add_to_pool` now provide the option to add the simulation day to the info map for the caller function. This functionality has not be made an option for logs, warnings, or errors - but easily could be.

A new utility method - `pad_temporal_data` has been added. This is the heart of the padding functionality. It is designed handle variables pulled out of the OutputManager's pool, and gives users options for the fill value and how values should be filled in after the last known value for a variable. Check the unit test for some good examples of how this function works.

### Test plan

Updated unit tests for the Output Manager and wrote new unit tests for `pad_temporal_data` - both suites are passing.

To test the end to end functionality, I wrote a JSON filter that only contained `Time.simulation_day` and checked that the simulation day in the info map matched the value recorded for the simulation day.

### Edit/Update
After speaking with Pooya, it was decided that defaulting to adding the simulation day to the info map of every variable added to the OM's pool was not a good idea at this time, given that it will expand the memory footprint and latency of recording variables when we are already short on memory and simulations take a long time. The data expansion capabilities will remain unchanged however.

### CODE REVIEW

File: RUFAS/util.py [hunk 1]
```diff
@@ -83,6 +84,82 @@ def flatten_keys_to_nested_structure(input_dict: Dict[str, Any]) -> Dict[str, An
...
+            keys "values" and optionally "info_maps".
+        fill_value : Any, default numpy.nan
+            Value that is used to pad the front of the data values, and optionally the back.
+        pad_tail_values : bool, default False
+            Whether tail values should be padded with the last known value from a data set, or with the fill value.
+
+        Returns
+        -------
+        dict[str, dict[str, list[Any]]]
+            The padded data, so that gaps in the data are filled in with the last know value or None.
```
Comment (JoeWaddell) [inline RUFAS/util.py]:
```suggestion
            The padded data, so that gaps in the data are filled in with the last known value or None.
```

File: RUFAS/util.py [hunk 2]
```diff
@@ -83,6 +84,82 @@ def flatten_keys_to_nested_structure(input_dict: Dict[str, Any]) -> Dict[str, An
 
         return nested_structure
 
+    @staticmethod
+    def pad_temporal_data(
```
Comment (JoeWaddell) [inline RUFAS/util.py]: Should it be clear that this is a daily padder in the name? If not possible in the name, then the description should probably make it clear(er) that it pads daily, since some information isn't reported on a daily interval.

Related, feel free to disregard: I'm assuming we shouldn't make this more complex and accept some interval argument (maybe in day increments), but it's a thought.

File: tests/test_util.py [hunk 3]
```diff
@@ -311,6 +312,195 @@ def test_flatten_keys_to_nested_structure_dict_w_list() -> None:
...
+    tail_pad: bool,
+    expected: dict[str, dict[str, list[Any]]],
+) -> None:
+    """Tests the utility method pad_temporal_data."""
+    actual = Utility.pad_temporal_data(data_to_pad, fill_value=fill_value, pad_tail_values=tail_pad)
+
+    assert actual == expected
+
+
+def test_pad_temporal_data_errors() -> None:
```
Comment (JoeWaddell) [inline tests/test_util.py]: Should you add a test/error for cases in which one is trying to pad a nonexistent field?

File: RUFAS/util.py [hunk 4]
```diff
@@ -83,6 +84,82 @@ def flatten_keys_to_nested_structure(input_dict: Dict[str, Any]) -> Dict[str, An
...
+        Pads data based on the simulation day(s) it was recorded on, relative to when other data was recorded.
+
+        Parameters
+        ----------
+        data_to_pad : dict[str, dict[str, list[Any]]]
+            The data to be padded. The top level key is a variable name, and points to a dictionary that contains the
+            keys "values" and optionally "info_maps".
+        fill_value : Any, default numpy.nan
+            Value that is used to pad the front of the data values, and optionally the back.
+        pad_tail_values : bool, default False
```
Comment (JoeWaddell) [inline RUFAS/util.py]: Per in person discussion, another flag to add `pad_gap_values` would be great, to allow the user to use the `fill_value` instead of the last known value in sections missing data (e.g. data only provided for days 1 and 100 of a 100 day simulation). Roughly follows behavior of `pad_tail_values` (maybe change to `pad_end_values`?)

Can be separate PR if too big for this one.

**Review by JoeWaddell — APPROVED** [review]
Looks great! Clear code, and good test coverage. A few name suggestions, and suggestion for new functionality/tests.

File: RUFAS/util.py (hunk 2 above)
Comment (Isgotkowitz) [reply RUFAS/util.py]: Made it clearer (I hope) in the description that data will be padded on a daily step for all values in a certain range.

> Related, feel free to disregard: I'm assuming we shouldn't make this more complex and accept some interval argument (maybe in day increments), but it's a thought.

I think it would be good to hold off on that functionality for right now, but that could definitely be feature added in the future.

Another feature that might become handy would be to pad "intermittently". I don't know a good way to describe this, so I'll just given an example:

Say `pad_temporal_data` gets variables x and y, with x being recorded on days 1, 5, and 10 and y being recorded on days 2, 7, and 9.

Currently, x and y will be padded so that they both have values for all days in the range [1, 10]. But a feature that pads "intermittently" would only add values to x and y so that they both have values for all days in the set {1, 2, 5, 7, 9, 10}.

File: tests/test_util.py (hunk 3 above)
Comment (Isgotkowitz) [reply tests/test_util.py]: What do you mean by a non-existent field?

Comment (JoeWaddell) [reply tests/test_util.py]: Ah yes, sorry for the lack of clarity: I was thinking of an empty dictionary.

File: RUFAS/util.py (hunk 4 above)
Comment (Isgotkowitz) [reply RUFAS/util.py]: Changed `pad_tail_values` to `fill_end_values`, and added the flag `fill_gap_values`. If `fill_gap_values` is true, then gaps between original values will be filled with the last know original value. If false, then gaps will be filled with the `fill_value`.

File: tests/test_util.py (hunk 3 above)
Comment (Isgotkowitz) [reply tests/test_util.py]: Yes! That should be checked against.

**Comment by Isgotkowitz** [comment]
After speaking with @PooyaHekmati, I am going to change the naming and docstrings surrounding this method to use the term "fill" instead of "pad" - this is because "padding" refers to adding data before or after existing data points, and "expanding" refers to adding data between existing data points. The functionality introduced does both, so I think "fill" is a good catch-all term. 

Tagging @JoeWaddell just to keep you updated.

File: tests/test_util.py (hunk 3 above)
Comment (Isgotkowitz) [reply tests/test_util.py]: Checked against!

File: RUFAS/util.py [hunk 5]
```diff
@@ -83,6 +84,98 @@ def flatten_keys_to_nested_structure(input_dict: Dict[str, Any]) -> Dict[str, An
...
+
+        all_simulation_days = []
+        for key, value in data_to_expand.items():
+            info_maps = value.get("info_maps")
+            if info_maps is None:
+                raise TypeError(f"Variable '{key}' has no info maps.")
+            if len(info_maps) != len(value["values"]):
+                raise ValueError(f"Variable '{key}' does not have matching number of values and info maps.")
+            if not all("simulation_day" in info_map.keys() for info_map in info_maps):
+                raise ValueError(f"Variable '{key}' does not have simulation day value in every info map.")
```
Comment (allisterakun) [inline RUFAS/util.py:144]: For these errors, do we need to add an error to the OM before raising the exception? Or maybe you'd think that's unnecessary?

File: RUFAS/util.py [hunk 6]
```diff
@@ -83,6 +84,98 @@ def flatten_keys_to_nested_structure(input_dict: Dict[str, Any]) -> Dict[str, An
...
+        filtered_simulation_days = sorted(set(all_simulation_days))
+        first_day = filtered_simulation_days[0]
+        last_day = filtered_simulation_days[-1]
+
+        expanded_data: dict[str, dict[str, list[Any]]] = {}
+        for key, data in data_to_expand.items():
+            expanded_variable_data: dict[str, list[Any]] = {"values": [], "info_maps": []}
+            original_units = data["info_maps"][0]["units"]
+            zipped_data = zip(data["values"], data["info_maps"])
+            indexed_data = {data[1]["simulation_day"]: data for data in zipped_data}
```
Comment (allisterakun) [inline RUFAS/util.py:156]: This is definitely not related to this PR, but what are your thoughts on enforcing some kind of data structure for all the add_variable data to follow?

**Review by allisterakun — APPROVED** [review]
Looks great, good job Ed!

**Review by allisterakun — APPROVED** [review]

File: RUFAS/util.py (hunk 5 above)
Comment (Isgotkowitz) [reply RUFAS/util.py:144]: Unfortunately we cannot import the OM into the Utility class because it will cause a circular import error.

File: RUFAS/util.py (hunk 6 above)
Comment (Isgotkowitz) [reply RUFAS/util.py:156]: I think that would be great, because all the parsing of dictionaries and zipping together data passed into the expand data method seems very fragile to me.

**Review by JoeWaddell — APPROVED** [review]
Great changes - method is useful, and implementation is easy to follow. Good test coverage. Easy approval!

Only final thoughts have to do with the naming of `fill_gap_values` and `fill_end_values`. I'm probably overthinking it, but a naive reader of the method might assume those are values to be used in gaps/the end, and not a boolean choice to use the `fill_value` or not. Likewise, `fill_gap_values` as a question would make me think that False wouldn't actually fill gaps (or the end).

All that said, I don't have a great suggestion for an alternative. `fill_gap_values` could be `use_fill_value_in_gaps` instead, but that's wordier and possibly still unclear (and would also require the flipping of the True/False logic, since here True would use the `fill_value`, and False would use the last known value). Some of this could maybe the decided in the subsequent PRs that use this functionality, e.g. https://github.com/RuminantFarmSystems/MASM/pull/1778

**Comment by Isgotkowitz** [comment]
> Great changes - method is useful, and implementation is easy to follow. Good test coverage. Easy approval!
> 
> Only final thoughts have to do with the naming of `fill_gap_values` and `fill_end_values`. I'm probably overthinking it, but a naive reader of the method might assume those are values to be used in gaps/the end, and not a boolean choice to use the `fill_value` or not. Likewise, `fill_gap_values` as a question would make me think that False wouldn't actually fill gaps (or the end).
> 
> All that said, I don't have a great suggestion for an alternative. `fill_gap_values` could be `use_fill_value_in_gaps` instead, but that's wordier and possibly still unclear (and would also require the flipping of the True/False logic, since here True would use the `fill_value`, and False would use the last known value). Some of this could maybe the decided in the subsequent PRs that use this functionality, e.g. #1778

Good suggestion, thank you @JoeWaddell. I will change the arg names to be more explicit and update the tests accordingly.

File: tests/test_input_manager.py
Comment (Isgotkowitz) [inline tests/test_input_manager.py:1]: I put this in here to fix a Flake8 violation that snuck onto main somehow.
